keps/sig-api-machinery/20190324-declatative-field-manager.md

```diff
@@ -0,0 +1,207 @@
+One of those actors is setting annotations for every service with external information.
+As the actor now only updates annotations not already owned by the `configManagement` fieldManager, applying them will cause the actor to own them.
+
+All actors now can coexist and manage their set of fields without interfering with each other.
+
+#### Kustomize
+
+In reference to the above workflow, it would be possible to override `options` for different Kustomize overlays and therefore separate ownership depending on the actor currently active.
+
+This means, updating single fields by using a Kustomize overlay (like updating images, labels or annotations) as part of a CI/CD pipeline could easily contain the fieldManager information without any changes to Kustomize (or similar tools) itself.
+
+## Design Details
+
+### Proposed API Change
+
+```go
+type ObjectMeta struct {
+...
+  // Options is a map of string keys and values for providing information about the object to the apiserver.
+  // This field is write-only, non-persisted and optional.
+  // It may only contain keys that are known to the current apiserver version.
+  // Sending a request with an invalid key/value results in the request being rejected.
+  // The apiserver validates the field against the `AllowedOptions` list.
+  // Options available through this field, that are also available as query options, should cause an error if both are set and differ.
+  Options map[string]string `json:"options,omitempty" protobuf:"bytes,18,rep,name=options"`
+...
+}
+
+// FieldManagerOption defines the ObjectMeta.Options key for setting the fieldManager.
+// It is only used by the apiserver on create, apply and update operations, to set the ManagedFields accordingly.
+// The value associated with this key, must be less than or 128 characters long, and only contain printable characters, as defined by https://golang.org/pkg/unicode/#IsPrint.
+//
+// If the option is unset or empty, the apiserver will default to the request user-agent.
+// If the request contains the fieldManager option, it acts like this option.
+// If both this option or the requests fieldManager option are set but not equal, the apply and non-apply operation will fail.
+//
+const FieldManagerOption = "fieldManager"
+
+// AllowedOptions defines the list of keys allowed for `ObjectMeta.Options`.
+var AllowedOptions = []string{
+  FieldManagerOption,
+}
+```
+
+### Test Plan
+
+e2e and integration tests should be added accordingly.
+
+### Graduation Criteria
+
+TBD
+
+### Upgrade / Downgrade Strategy
+
+No fields get persisted.
+When the field is not set due to an old client, the current Server Side Apply defaulting is the fallback.
+
+The apiserver should always reject unknown keys in the options field to provide the user with clear information
+and prevent unexpected behavior cause by ignored/discarded options.
```

Comment:
If option is discarded, there wouldn't be unexpected behavior :-)
Better rephrase this.